syslog-ng/python-modules/webhook/source.py

```diff
@@ -0,0 +1,139 @@
+#############################################################################
+# Copyright (c) 2024 László Várady
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License version 2 as published
+# by the Free Software Foundation, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+#
+# As an additional exemption you are allowed to compile & link against the
+# OpenSSL libraries as published by the OpenSSL project. See the file
+# COPYING for details.
+#
+#############################################################################
+
+from syslogng import LogSource, LogMessage
+
+import logging
+import asyncio
+import threading
+import tornado
+import ssl
+import signal
+from typing import Any
+
+signal.signal(signal.SIGINT, signal.SIG_IGN)
+signal.signal(signal.SIGTERM, signal.SIG_IGN)
+
+class Handler(tornado.web.RequestHandler):
+    def initialize(self, source) -> None:
+        self.source = source
+
+    @tornado.web.authenticated
+    async def post(self) -> None:
+        # racy, but the client should retry
+        if self.source.suspended.is_set():
+            self.set_status(503)
+            await self.finish({"status": "flow-controlled"})
+            return
+
+        self.source.post_message(LogMessage(self.request.body))
+        await self.finish({"status": "received"})
+
+    def set_default_headers(self) -> None:
+        self.set_header("Server", "syslog-ng");
+
+    def get_current_user(self):
+        if not self.source.auth_token:
+            return True
+
+        token = self.request.headers.get("Authorization", "").split(" ")
+        if len(token) != 2:
+            return False
+
+        token = token[1]
+        return self.source.auth_token == token
+
+    def write_error(self, status_code: int, **kwargs: Any) -> None:
+        self.set_status(status_code)
+
+class HTTPSource(LogSource):
+    def init(self, options: dict[str, Any]) -> bool:
+        self.logger = logging.getLogger("http")
+        self.set_transport_name("http")
+        if not self.init_options(options):
+            return False
+
+        self.ssl_ctx = None
+        if self.tls_key_file:
+            self.ssl_ctx = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
+            self.ssl_ctx.load_cert_chain(self.tls_cert_file, self.tls_key_file)
```

Comment:
As far as I can understand, without CAs loaded we do not verify the client's cert. Do I understand right? Don't we want to do so?

For reference:
 * https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_default_certs
 * https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_verify_locations

Reply:
We don't want this right now.

Reply:
I'm adding this, just in case anyone needs it.